pkg/ingester/label_names_and_values.go

```diff
@@ -167,6 +186,30 @@ func labelValuesCardinality(
 	return nil
 }
 
+func computeLabelValuesSeriesCount(
+	ctx context.Context,
+	lbName string,
+	lbValues []string,
+	idxReader tsdb.IndexReader,
+	matchedPostingsCloner *index.PostingsCloner,
+	countCh chan<- labelValueCountResult,
+) {
```

Comment:
I think that usually the pattern is a function that gives you a channel with the results, not a function that accepts a channel to write the results into.

I'd:
- make this function return a channel
- call this function synchronously, and this function would:
```go
chan := make(chan labelValueCountResult, len(lbValues))
wg := sync.WaitGroup{}
wg.Add(len(lbValues))
go concurrency.ForEachJob(...)
go func() {
    wg.Wait()
    close(chan)
}()
return chan
```

Reply:
Maybe `concurrency.ForEachJob` isn't such a good fit here after all.

Reply:
When I've used it previously there were matching "job" and "result" slices created: https://github.com/grafana/mimir/blob/main/pkg/querier/tenantfederation/merge_exemplar_queryable.go#L138-L187

Reply:
And what if we simply use a separated channel to signal completion? This way we would ensure that none of the producers panics while sending the data. 

```go
resultCh , doneCh := computeLabelValuesSeriesCount(ctx, lbName, allLblValues, idxReader, mpc)

running := true
for running {
  select {
    case res := <-resultCh:
       ...
     case <-doneCh:
       running = false
  }
}
```

Reply:
@56quarters in that example we're just paralleling the work, but here we also want to stream out the results while we're getting them.

@ortuman now we moved the problem to signalling the `doneCh`, which is the same problem but on a different channel, right? :D

Reply:
> @ortuman now we moved the problem to signalling the doneCh, which is the same problem but on a different channel, right? :D

The doneCh would by closed by the goroutine in charge of waiting for jobs completion. This would avoid the panic that you mentioned above. May I be missing something?

```go
resultCh = make(chan labelValueCountResult, len(lbValues))
doneCh = make(chan struct{})

go func() {
	// Failing job error will be sent over the countCh channel, so no need to handle it here.
	_ = concurrency.ForEachJob(ctx, len(lbValues), maxConcurrency, func(ctx context.Context, idx int) error {
		lbValue := lbValues[idx]

		seriesCount, err := countLabelValueSeries(ctx, lbName, lbValue, idxReader, matchedPostingsCloner)

		resultCh <- labelValueCountResult{
			val:   lbValue,
			count: seriesCount,
			err:   err,
		}
		return err
	})
	close(doneCh)
}()
return
```

Reply:
This code doesn't ensure that all results from `resultCh` are read.

Consider there's only one thread scheduling goroutines.
It writes all the results into `resultCh`. Then it closes `doneCh`. Then it goes back to executing your for loop, and randomly selects from `resultCh` and `doneCh`.

Reply:
Oh, I see your point.   🤦‍♂️ 

I wonder if we could simply close `resultCh` only when `concurrency.ForEachJob` returns no error. That would imply that all results have been written to the channel. Otherwise we leave it opened to avoid the condition you exposed above. Eventually the consumer will get the error through the result channel and return immediately.

Reply:
You can check on the result of `concurrency.ForEachJob`, and if `err == nil` you can safely close `resultCh` (and rename `s/doneCh/failedCh/`).

However, notice that we're writing more lines of code to workaround the behavior of `concurrency.ForEachJob` than the lines of code that method has. 

This is my suggestion:
```go
func computeLabelValuesSeriesCount(ctx context.Context, lbName string, lbValues []string, idxReader tsdb.IndexReader, matchedPostingsCloner *index.PostingsCloner) chan labelValueCountResult {
	countCh := make(chan labelValueCountResult)

	wg := sync.WaitGroup{}
	wg.Add(len(lbValues))

	indexes := atomic.NewInt64(-1)
	for ix := 0; ix < maxConcurrency; ix++ {
		go func() {
			for {
				idx := int(indexes.Inc())
				if idx >= len(lbValues) {
					return
				}
				seriesCount, err := countLabelValueSeries(ctx, lbName, lbValues[idx], idxReader, matchedPostingsCloner)
				countCh <- labelValueCountResult{
					val:   lbValues[idx],
					count: seriesCount,
					err:   err,
				}
				wg.Done()
			}
		}()
	}
	go func() {
		wg.Wait()
		close(countCh)
	}()

	return countCh
}
```

Reply:
Sweet. I'll borrow that. 👍

Reply:
done @ befacb1b55eb27cc9fe50cdaa71175705b4951f4